aten/src/ATen/cuda/CUDAContext.h

```diff
@@ -36,6 +36,65 @@ It is expected that the modules whose functions compose this interface will
 manage their own state. There is only a single CUDA context/state.
 */
 
+struct CAFFE2_API CUDAP2PState {
+  CUDAP2PState(int64_t src, int64_t target) 
+  : src_device(src), target_device(target) {
+    auto num_gpus = c10::cuda::device_count();
+    AT_ASSERT(src_device >= 0 && src_device < num_gpus);
+    AT_ASSERT(target_device >= 0 && target_device < num_gpus);
+    // if same device, no need for enabling
+    if(src_device == target_device) {
+      p2p_enable_status = true;
+    // if existing connection there, no need for enabling
+    } else if (!(p2p_enable_status)) {
+      auto current_device = c10::cuda::current_device();
+      c10::cuda::set_device(src_device);
+
+      int canAccess;
+      AT_CUDA_CHECK(cudaDeviceCanAccessPeer(&canAccess, src_device, target_device));
+      if (canAccess) {
+        cudaError_t err = cudaDeviceEnablePeerAccess(target_device, 0);
+        if (err == cudaErrorPeerAccessAlreadyEnabled || err == cudaSuccess) {
```

Comment:
This looks very dodgy. Suppose two threads create CUDAP2PState(0,1). They will both attempt to enable the P2P access. Fine, you've written code to work around it. Then on destruction of the first state, peer access will be disabled, EVEN if another thread thinks it still wants it.